### ISSUE TITLE
GameGuru Max - Request - Enhancement - Climbable Surfaces (Objects/Edges)

### ISSUE THREAD

**Issue by Lemuripest93** [TheGameCreators/GameGuruRepo#3421]
Title speaks for itself, having climbable surfaces, such as jumping and getting to grip unto the edge of a roof, or platform - or anything really - would be super useful.

This would mean you could have a new layer of versatility in terms of movement and most importantly - much more interesting level design if someone should desire to add a bit of platforming to their game, or maybe alternate paths to an objective? secrets? this could open up new paths across all the available genres in GGM (when available) as far as i see it.

How does it work? pretty much like the gif below, you jump, you hold the spacebar after jumping, you climb, easy.

This could be a behavior, essentially you pick an object, set it to 'Physics Off' and add the "Climbable Surface" behavior to it, and voilà you got yourself a handy climbable surface.

Alternatively this could be a player setting, so basically when you tick off the "can climb" box under the player settings, any object in the players reach becomes climbable.

Both of these options are pretty much usable, for instance the first option would be good for a game with only a few platforming segments whilst the latter option would be good for a game where the player and level design is more flexible.

![ggmpreyclimb](https://user-images.githubusercontent.com/104637906/178122199-a183f4cd-043b-4f5c-8f19-c795fbb1b71d.gif)
(footage from the excellent Prey from 2017)

**Comment by Necrym59** [comment]
Already written
[climber_v23.zip](https://github.com/TheGameCreators/GameGuruRepo/files/9077978/climber_v23.zip)

**Comment by Lemuripest93** [comment]
> Already written [climber_v23.zip](https://github.com/TheGameCreators/GameGuruRepo/files/9077978/climber_v23.zip)

Pretty decent but if i had to make a criticism its that it functions a little too much like a ladder, the player shouldn't automatically climb, but rather have to hold down a button to climb, basically like i said above - space to jump, then hold down to climb, i think if you made this one change, you'd have it spot on.

Otherwise this basically functions like the first option in my main post - thanks Necrym 👍

**Comment by Necrym59** [comment]
Ok ill have a look at putting that option in for jump/grab and climbSent on the go with Vodafone
-------- Original message --------From: Lemuripest93 ***@***.***> Date: 10/7/22  9:32 am  (GMT+10:00) To: TheGameCreators/GameGuruRepo ***@***.***> Cc: Necrym59 ***@***.***>, Comment ***@***.***> Subject: Re: [TheGameCreators/GameGuruRepo] GameGuru Max - Request - Enhancement - Climbable Surfaces (Objects/Edges) (Issue #3421) 



Already written climber_v23.zip



Pretty decent but if i had to make a criticism its that it functions a little too much like a ladder, the player shouldn't automatically climb, but rather have to hold down a button to climb, basically like i said above - space to jump, then hold down to climb, i think if you made this one change, you'd have it spot on.

Otherwise this basically functions like the first option in my main post - thanks Necrym 👍

**Comment by Necrym59** [comment]
Which button do you think is better for a grab - a key or mouse 2

**Comment by fearlesswee** [comment]
I think most games call this "clambering", "vaulting", or "ledge grabbing", where the player can grab onto and pull themselves up/over any ledge/obstacle by pressing the jump key which helps with jumping over larger objects, quickly mantling over cover in the way, or crossing a gap if they just barely miss the jump. You can see it being used to great effect in the recent DOOM games (2016 and Eternal).

**Comment by Lemuripest93** [comment]
> Which button do you think is better for a grab - a key or mouse 2

The traditional way this is done is by holding down the button that jumps, so in this case it should be the spacebar.

**Comment by Necrym59** [comment]
Might be i wil;l look at making a new script for these features and leave the existing one alone that way we can have horizontal/vertical/angled gripper objects for these uses, and can be used in tandem with the standard climber object if required.

**Comment by Tam51** [comment]
I also think it would be great if we could climb left and right when your on say a rock cliff face.

**Comment by Lemuripest93** [comment]
> Might be i wil;l look at making a new script for these features and leave the existing one alone that way we can have horizontal/vertical/angled gripper objects for these uses, and can be used in tandem with the standard climber object if required.

Sounds good, i can actually see myself using the climbing script you posted in a project, i've been playing around with a few different ideas for how i could use it. Great that you're open to making a few different versions for different climbing angles, looking forward to future updates on that front.

There is actually a new change i'd like though, could you make it so when you aim left or right with the mouse, you jump off towards that direction? currently you jump off in the opposite direction you're looking and it feels a bit awkward.

**Comment by Lemuripest93** [comment]
Wanted to update this post a little with a new point of reference, 'Prey' is a great game and all but if there is one game recently that truly shows off the potential of this feature, it's 'Cyberpunk 2077' (as it got patched into a pretty solid game.)

Recorded some footage below in which i kill off some thugs by using some creative means instead of a direct approach.

Video Demonstration: https://www.youtube.com/watch?v=KRNR0mTcsIE

Basically the whole point of this feature is to allow people to make creative quest design and combat encounters, as well as adding a bit of parkour to their world.
